AmigaOS4 (or general): "Ressurect in Asylum" didn't work

Find out (At least on AmigaOS4), that if we go for example left/left in the first level to the table with a pumpkin head, and press "f" there so go to a new level and then die there, we have a question: resurrect or resurrect in Asylum. If I choose pure resurrect, then all fine, the game starts from the beginning, if we choose to resurrect in Asylum, then nothing changes. In console I have :

"Warn failed: "false" ()
in unknown fucntion in code/Projects/Eldritch/src/eldritchgame.cpp at line 275.

@all
Checked latest sources, when tried to build also have that:

```
[  0%] Building CXX object Libraries/Core/CMakeFiles/Core.dir/src/allocator.cpp.obj
In file included from /amiga/Eldritch-master/code/Libraries/Core/src/allocator.cpp:4:
/amiga/Eldritch-master/code/Libraries/Core/src/idatastream.h:54:26: error: ‘uint8’ has not been declared
   54 |  inline void WriteUInt8( uint8 i ) const { Write( 1, &i ); }
      |                          ^~~~~
/amiga/Eldritch-master/code/Libraries/Core/src/idatastream.h:55:27: error: ‘uint16’ has not been declared
   55 |  inline void WriteUInt16( uint16 i ) const { littleBigEndian( &i ); Write( 2, &i ); }
      |                           ^~~~~~
/amiga/Eldritch-master/code/Libraries/Core/src/idatastream.h:56:27: error: ‘uint32’ has not been declared
   56 |  inline void WriteUInt32( uint32 i ) const { littleBigEndian( &i ); Write( 4, &i ); }
      |                           ^~~~~~
/amiga/Eldritch-master/code/Libraries/Core/src/idatastream.h:57:25: error: ‘int8’ has not been declared
   57 |  inline void WriteInt8( int8 i ) const { Write( 1, &i ); }
      |                         ^~~~
/amiga/Eldritch-master/code/Libraries/Core/src/idatastream.h:58:26: error: ‘int16’ has not been declared
   58 |  inline void WriteInt16( int16 i ) const { littleBigEndian( &i ); Write( 2, &i ); }
      |                          ^~~~~
/amiga/Eldritch-master/code/Libraries/Core/src/idatastream.h:59:26: error: ‘int32’ has not been declared
   59 |  inline void WriteInt32( int32 i ) const { littleBigEndian( &i ); Write( 4, &i ); }
```

Probabaly that because there all ints/units etc now are c_uint8, etc ?

@rajdakin 
Tried fix8, yeah, this uint/int issues gone.
Next one 

```
code/Libraries/Core/src/idatastream.h:144:20: error: ‘i’ was not declared in this scope
  144 |   littleBigEndian(&i);
```

As far as I can see should be ```f```

Next error is in "allocator.h" and "allocator.cpp", also gl2texture and gl2cubesphera about ```error: ‘uintptr_t’ does not name a type```. I fixed that by adding ```#include <stdint.h>``` to allocator.h at top. And the same ```#include <stdint.h>``` at top of idatastream.h

Another few errors is about few unimplemented functions in fileutil.cpp and timedate.cpp _ timedate.h. The all can be fixed by changing all ```#elif BUILD_LINUX || BUILD_MAC``` on ```#elif BUILD_LINUX || BUILD_MAC || __amigaos4__```


Another error in meshfactory.cpp:1465:27 :
```
code/Libraries/3D/src/meshfactory.cpp:1465:27: error: ‘struct SCompiledMeshHeader’ has no member named ‘m_NumCollisionTris’; did you mean ‘m_PADDING_NumCollisionTris’?
 1465 |  littleBigEndian( &Header.m_NumCollisionTris );
      |                           ^~~~~~~~~~~~~~~~~
```

That seems because in v8 fixes about Pading didn't fits ?

Thanks, tested new v8:

allocator.h still need ```#include <cstdint>``` too

and then fileutil.cpp also need those changes with adding ```|| __amigaos4__``` on the lines 386 and 468

After fixing that, i were able to build all source code. Thorugh i fail on linking with:

```
../../Libraries/SoLoud/libSoLoud.a(soloud.cpp.obj): In function `_ZN6SoLoud6Soloud4initEjjjjj':
soloud.cpp:(.text+0x56c): undefined reference to `_ZN6SoLoud8sdl_initEPNS_6SoloudEjjjj'
```

But that i need to check on my side, probabaly some missing define somewhere.. Maybe i need to add some additional flag to Cmake line now ?

> I don't think allocator.h needs this include, what is the error there? (I've built using -D__amigaos4__ and I didn't get any error...)

Same as when there is no include of cstdint in other ones: 

```
user@DESKTOP-3NFAB0O /amiga/Eldritch-fix8/code/build
$ make Eld
Consolidate compiler generated dependencies of target UI
[  5%] Built target UI
Consolidate compiler generated dependencies of target Core
[  5%] Building CXX object Libraries/Core/CMakeFiles/Core.dir/src/allocator.cpp.obj
[  6%] Building CXX object Libraries/Core/CMakeFiles/Core.dir/src/allocatorchunk.cpp.obj
/amiga/Eldritch-fix8/code/Libraries/Core/src/allocatorchunk.cpp: In member function ‘void AllocatorChunk::SplitToAlignment(AllocatorChunk::SAllocatorBlock*, uint)’:
/amiga/Eldritch-fix8/code/Libraries/Core/src/allocatorchunk.cpp:293:58: error: ‘uintptr_t’ does not name a type
  293 |  const uint AlignPad    = Alignment - ( reinterpret_cast<uintptr_t>( reinterpret_cast<byte*>( BlockToSplit ) + ( Allocator::m_BlockHeaderSize << 1 ) ) % Alignment );
      |                                                          ^~~~~~~~~
/amiga/Eldritch-fix8/code/Libraries/Core/src/allocatorchunk.cpp: In member function ‘bool AllocatorChunk::CanSplitToAlignment(const AllocatorChunk::SAllocatorBlock*, uint, uint) const’:
/amiga/Eldritch-fix8/code/Libraries/Core/src/allocatorchunk.cpp:324:51: error: ‘uintptr_t’ does not name a type
  324 |  uint AlignPad   = Alignment - ( reinterpret_cast<uintptr_t>( reinterpret_cast<const byte*>( BlockToSplit ) + ( Allocator::m_BlockHeaderSize << 1 ) ) % Alignment );
      |                                                   ^~~~~~~~~
/amiga/Eldritch-fix8/code/Libraries/Core/src/allocatorchunk.cpp: In member function ‘bool AllocatorChunk::IsAligned(const AllocatorChunk::SAllocatorBlock*, uint) const’:
/amiga/Eldritch-fix8/code/Libraries/Core/src/allocatorchunk.cpp:358:28: error: ‘uintptr_t’ does not name a type
  358 |  return ( reinterpret_cast<uintptr_t>( reinterpret_cast<const byte*>( pBlock ) + Allocator::m_BlockHeaderSize ) % Alignment ) == 0;
      |                            ^~~~~~~~~
make[3]: *** [Libraries/Core/CMakeFiles/Core.dir/build.make:132: Libraries/Core/CMakeFiles/Core.dir/src/allocatorchunk.cpp.obj] Error 1
make[2]: *** [CMakeFiles/Makefile2:350: Libraries/Core/CMakeFiles/Core.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:629: Projects/Eld/CMakeFiles/Eld.dir/rule] Error 2
make: *** [Makefile:254: Eld] Error 2
```

As for the sound: dunno, I run CMAKE as before, like this for the moment:

```
cmake \
-DCMAKE_SYSTEM_NAME=Generic \
-DCMAKE_SYSTEM_VERSION=1 \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_C_COMPILER="/usr/local/amiga/bin/ppc-amigaos-gcc" \
-DCMAKE_CXX_COMPILER="/usr/local/amiga/bin/ppc-amigaos-g++" \
-DCMAKE_CXX_FLAGS="-D__USE_AMIGAOS_NAMESPACE__" \
-DCMAKE_LINKER="/usr/local/amiga/bin/ppc-amigaos-ld" \
-DCMAKE_AR="/usr/local/amiga/bin/ppc-amigaos-gcc-ar" \
-DCMAKE_RANLIB="/usr/local/amiga/bin/ppc-amigaos-gcc-ranlib" \
-DCMAKE_FIND_ROOT_PATH="/usr/local/amiga/ppc-amigaos/" \
-DGLES=ON \
..
```

I probably better to use SDL2 sound backend, should i add something like ```-DWITH_SDL=ON``` ?

--- a/Code/Libraries/3D/src/surface.cpp
+++ b/Code/Libraries/3D/src/surface.cpp
@@ -111,6 +111,24 @@ int Surface::LoadBMP( const IDataStream& Stream )
 
 	Stream.Read( sizeof( SBitmapFileHeader ), &BMPFileHeader );
 	Stream.Read( sizeof( SBitmapInfoHeader ), &BMPInfoHeader );
+#ifdef __amigaos4__
+	littleBigEndian( &BMPFileHeader.m_Type );
+	littleBigEndian( &BMPFileHeader.m_Size );
+	littleBigEndian( &BMPFileHeader.m_Reserved1 );
+	littleBigEndian( &BMPFileHeader.m_Reserved2 );
+	littleBigEndian( &BMPFileHeader.m_OffsetBits );
+	littleBigEndian( &BMPInfoHeader.m_Size );
+	littleBigEndian( &BMPInfoHeader.m_Width );
+	littleBigEndian( &BMPInfoHeader.m_Height );
+	littleBigEndian( &BMPInfoHeader.m_Planes );
+	littleBigEndian( &BMPInfoHeader.m_BitCount );
+	littleBigEndian( &BMPInfoHeader.m_Compression );
+	littleBigEndian( &BMPInfoHeader.m_SizeImage );
+	littleBigEndian( &BMPInfoHeader.m_PixelsPerMeterX );
+	littleBigEndian( &BMPInfoHeader.m_PixelsPerMeterY );
+	littleBigEndian( &BMPInfoHeader.m_ColorUsed );
+	littleBigEndian( &BMPInfoHeader.m_ColorImportant );
+#endif
 
 	// Support extensions by reading whatever is available into header
 	SBitmapInfoHeaderExt	BMPInfoHeaderExt = { 0 };
@@ -121,6 +139,23 @@ int Surface::LoadBMP( const IDataStream& Stream )
 	if( Remaining > 0 )
 	{
 		Stream.Read( Min( SizeOfExt, Remaining ), &BMPInfoHeaderExt );
+#ifdef __amigaos4__
+	littleBigEndian( &BMPInfoHeaderExt.m_RedMask );
+	littleBigEndian( &BMPInfoHeaderExt.m_GreenMask );
+	littleBigEndian( &BMPInfoHeaderExt.m_BlueMask );
+	littleBigEndian( &BMPInfoHeaderExt.m_AlphaMask );
+	littleBigEndian( &BMPInfoHeaderExt.m_ColorSpaceType );
+	littleBigEndian( &BMPInfoHeaderExt.m_EndpointsX );
+	littleBigEndian( &BMPInfoHeaderExt.m_EndpointsY );
+	littleBigEndian( &BMPInfoHeaderExt.m_EndpointsZ );
+	littleBigEndian( &BMPInfoHeaderExt.m_GammaRed );
+	littleBigEndian( &BMPInfoHeaderExt.m_GammaGreen );
+	littleBigEndian( &BMPInfoHeaderExt.m_GammaBlue );
+	littleBigEndian( &BMPInfoHeaderExt.m_Intent );
+	littleBigEndian( &BMPInfoHeaderExt.m_ProfileData );
+	littleBigEndian( &BMPInfoHeaderExt.m_ProfileSize );
+	littleBigEndian( &BMPInfoHeaderExt.m_Reserved );
+#endif
 		Remaining -= SizeOfExt;
 		NeedsSwizzle = true;	// HACKHACK, see below
 
@@ -352,10 +387,19 @@ void Surface::SaveBMP( const IDataStream& Stream )
 	Header.m_Type		= 'MB';
 	Header.m_Size		= static_cast<uint>( 54 + Size );
 	Header.m_OffsetBits	= 54;
+#ifdef __amigaos4__
+	littleBigEndian( &m_BitmapInfo.m_Header.m_SizeImage );
+	littleBigEndian( &Header.m_Type );
+	littleBigEndian( &Header.m_Size );
+	littleBigEndian( &Header.m_OffsetBits );
+#endif
 
 	Stream.Write( sizeof( SBitmapFileHeader ), &Header );
 	Stream.Write( sizeof( SBitmapInfoHeader ), &m_BitmapInfo.m_Header );
 	Stream.Write( Size, m_pPixels );
+#ifdef __amigaos4__
+	littleBigEndian( &m_BitmapInfo.m_Header.m_SizeImage );
+#endif
 }
 
 // This assumes that both Surfaces are the same format and bit depth

--- a/Code/Libraries/SoLoud/src/core/soloud_file.cpp
+++ b/Code/Libraries/SoLoud/src/core/soloud_file.cpp
@@ -43,6 +43,9 @@ namespace SoLoud
 	{
 		unsigned short d = 0;
 		read((unsigned char*)&d, 2);
+#ifdef __amigaos4__
+		d = (d >> 8) | (d << 8);
+#endif
 		return d;
 	}
 
@@ -50,6 +53,9 @@ namespace SoLoud
 	{
 		unsigned int d = 0;
 		read((unsigned char*)&d, 4);
+#ifdef __amigaos4__
+		d = (d >> 24) | ((d >> 8) & 0xFF00) | ((d << 8) & 0xFF0000) | (d << 24);
+#endif
 		return d;
 	}
 

